### PR TITLE
internal links

### LINKED ISSUE #281: Add support for internal links
- [x] Markdown editor plugin to choose linked document (should insert document ID into Markdown)
- [x] Markdown renderer plugin to correctly render those links (python)
- [ ] ~~Markdown renderer plugin to correctly render those links (javascript)~~ (-> #293)

### PR DESCRIPTION
**Pull request by Nef10**
fix #281

### CODE REVIEW

File: _1327/documents/markdown_internal_links.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+MINUTES_RE = r'\[([^\[]+)\]\(minutes:(\d+)\)'
+POLLS_RE = r'\[([^\[]+)\]\(poll:(\d+)\)'
+INFORMATION_DOCUMENT_RE = r'\[([^\[]+)\]\(information_document:(\d+)\)'
+
+
+class InternalLinkPattern (LinkPattern):
+
+	def handleMatch(self, m):
+		title = m.group(2)
```
Comment (Bartzi) [inline _1327/documents/markdown_internal_links.py]: how about using named groups here?

File: _1327/documents/markdown_internal_links.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		document = MinutesDocument.objects.get(id=id)
+		if document:
+			return '/minutes/' + document.url_title
+		return ''
+
+
+class InternalLinkPollsPattern (InternalLinkPattern):
+
+	def url(self, id):
+		document = Poll.objects.get(id=id)
```
Comment (Bartzi) [inline _1327/documents/markdown_internal_links.py]: could you rename this to `poll`?

File: _1327/documents/templates/editor.html [hunk 3]
```diff
@@ -196,6 +231,38 @@
...
+							setButton.on("click", function (event) {
+								modalItem.modal('hide');
+								// find the right spot to insert the text and create the correct markdown
+								var markdown = $('#selectInternalLinkModal').val();
+								if (markdown !== null) {
+									linkText = markdown.match(/\[(.*?)\]/)[1]; // we want the text inside the [ ]
+									selected = e.getSelection();
+									e.replaceSelection(markdown);
+									cursor = selected.start;
+									e.setSelection(cursor + 1, cursor + linkText.length + 1) // +1 because we do not want to select the starting [
```
Comment (janno42) [inline _1327/documents/templates/editor.html]: in firefox the selection seems to have no effect.
in chrome the text gets selected, but the edit field loses the focus once the modal is closed and the selection is not there anymore.

File: _1327/documents/markdown_internal_links.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import markdown
+from markdown.inlinepatterns import LinkPattern
+
+from _1327.information_pages.models import InformationDocument
+from _1327.minutes.models import MinutesDocument
+from _1327.polls.models import Poll
+
+MINUTES_RE = r'\[(?P<title>[^\[]+)\]\(minutes:(?P<id>\d+)\)'
+POLLS_RE = r'\[(?P<title>[^\[]+)\]\(poll:(?P<id>\d+)\)'
+INFORMATION_DOCUMENT_RE = r'\[(?P<title>[^\[]+)\]\(information_document:(?P<id>\d+)\)'
```
Comment (invliD) [inline _1327/documents/markdown_internal_links.py]: These should come from the actual packages implementing the type.

File: _1327/documents/markdown_internal_links.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+			return '/polls/' + str(poll.id) + '/results'
+		return ''
+
+
+class InternalLinkInformationDocumentPattern (InternalLinkPattern):
+
+	def url(self, id):
+		document = InformationDocument.objects.get(id=id)
+		if document:
+			return '/' + document.url_title
```
Comment (invliD) [inline _1327/documents/markdown_internal_links.py]: Use the Django url helpers to generate these URLs.

File: _1327/documents/templates/editor.html [hunk 6]
```diff
@@ -73,6 +73,41 @@
...
 	</div>
 
+	<div class="modal fade ontop" id="selectInternalLinkDestination" tabindex="-1" role="dialog" aria-hidden="true">
+		<div class="modal-dialog">
+			<div class="modal-content">
+				<div class="modal-header">
+					<button type="button" class="close" data-dismiss="modal" aria-label="Close">
+						<span aria-hidden="true">&times;</span>
+					</button>
+					<div class="modal-title">{% trans "Select link target:" %}</div>
```
Comment (invliD) [inline _1327/documents/templates/editor.html]: No colons in titles.

**Comment by Nef10** [comment]
@invliD @janno42 as I am not available this and the next week feel free to work on this PR

File: _1327/documents/templates/editor.html [hunk 7]
```diff
@@ -73,6 +73,41 @@
...
+				</div>
+				<div class="modal-body">
+					<select class="form-control" id="selectInternalLinkModal">
```
Comment (janno42) [inline _1327/documents/templates/editor.html]: apply select2

**Comment by janno42** [comment]
all comments addressed

File: _1327/documents/markdown_internal_links.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		el = markdown.util.etree.Element("a")
+		el.set('href', self.url(m.group('id')))
+		el.text = markdown.util.AtomicString(m.group('title'))
+		return el
+
+	def url(id):
+		raise NotImplementedError
+
+
+class InternalLinkMinutesPattern (InternalLinkPattern):
```
Comment (invliD) [inline _1327/documents/markdown_internal_links.py]: These should be in the respective Django app, not in documents. And wouldn't it be nice to implement the interface in the actual model?

File: _1327/static/less/1327.less [hunk 9]
```diff
@@ -361,3 +361,10 @@ table.diff > thead th:nth-of-type(4) { width: 45%; }
 .alert.alert-warning {
 	color: @gray-darker;
 }
+
+// fix for select2 in modal
+.select2-container {
+	width: 100% !important;
+	padding: 0;
+	z-index: 99999;
+}
```
Comment (invliD) [inline _1327/static/less/1327.less]: Newline.

Comment (janno42) [inline _1327/static/less/1327.less]: was fixed before you commented ;)

File: _1327/documents/templates/editor.html [hunk 10]
```diff
@@ -2,231 +2,305 @@
 {% load i18n %}
```
Comment (invliD) [inline _1327/documents/templates/editor.html]: It's really hard to see what you changed in this file. Please separate formatting and code changes in different commits or pull requests.

File: _1327/static/less/1327.less (hunk 9 above)
Comment (invliD) [inline _1327/static/less/1327.less]: Not according to GitHub. It still shows the missing newline here.

File: _1327/documents/templates/editor.html [hunk 11]
```diff
@@ -2,231 +2,305 @@
...
+							modalItem.modal('hide');
+							// find the right spot to insert the text and create the correct markdown
+							var markdown = $('#selectInternalLinkModal').val();
+							if (markdown !== null) {
+								linkText = markdown.match(/\[(.*?)\]/)[1]; // we want the text inside the [ ]
+								selected = e.getSelection();
+								e.replaceSelection(markdown);
+								cursor = selected.start;
+								e.setSelection(cursor + 1, cursor + linkText.length + 1) // +1 because we do not want to select the starting [
+								$("#id_text").focus()
```
Comment (janno42) [inline _1327/documents/templates/editor.html]: this line was added

File: _1327/information_pages/views.py [hunk 12]
```diff
@@ -51,6 +54,11 @@ def edit(request, title, new_autosaved_pages=None):
 		messages.success(request, _("Successfully saved changes"))
 		return HttpResponseRedirect(reverse('information_pages:view_information', args=[document.url_title]))
 	else:
+
+		minutes = get_objects_for_user(request.user, MinutesDocument.VIEW_PERMISSION_NAME, klass=MinutesDocument)
+		information_documents = get_objects_for_user(request.user, InformationDocument.VIEW_PERMISSION_NAME, klass=InformationDocument)
+		polls = get_objects_for_user(request.user, Poll.VIEW_PERMISSION_NAME, klass=Poll)
+
```
Comment (invliD) [inline _1327/information_pages/views.py]: It seems like a bad idea to load these all the time. Can't verify the JavaScript until you fix the diff, however. I'm not gonna read through that entire file to figure out what you actually changed.

File: _1327/documents/templates/editor.html [hunk 13]
```diff
@@ -2,231 +2,305 @@
...
-		save();
+		setTimeout(function() { save(); }, 10000);
+	}
+	save();
+
+	$("#selectInternalLinkModal").select2({
+		language: "{{ LANGUAGE_CODE }}",
+		placeholder: "{% trans "Please select..." %}",
+		minimumResultsForSearch: 3
+	});
```
Comment (janno42) [inline _1327/documents/templates/editor.html]: these 5 lines were added

File: _1327/documents/templates/editor.html (hunk 10 above)
Comment (janno42) [inline _1327/documents/templates/editor.html]: i commented on the two changed parts

Comment (invliD) [inline _1327/documents/templates/editor.html]: That's not what I asked. Please separate it into multiple commits so someone else will find the actual change as well.

File: _1327/documents/templates/editor.html [hunk 14]
```diff
@@ -2,231 +2,305 @@
...
+			});
 		}
-		save();
+		setTimeout(function() { save(); }, 10000);
+	}
+	save();
+
+	$("#selectInternalLinkModal").select2({
+		language: "{{ LANGUAGE_CODE }}",
+		placeholder: "{% trans "Please select..." %}",
```
Comment (invliD) [inline _1327/documents/templates/editor.html]: Please use single quotes for the inner quotes.

**Comment by invliD** [comment]
That's better. Thanks @janno42. Please see my other comments.

**Comment by janno42** [comment]
:+1:
